File: madgoshka/Adrian/Madgo/TAC.hs
-- What:
-- This layer translates expressions that are composed of several smaller ones
-- to three address code.
--
-- Why:
-- Because every variable must be freed. When we create an variable like
-- this: `var lol: c#IntFast32 = c#IntFast32(0) + c#IntFast32(1)`
-- we have no variables that point to `c#IntFast32(0)` and `c#IntFast32(1)`.
-- Also these transformations will help us to improve output code.
--
-- Example:
-- input:  var someVariable: c#IntFast8 = c#IntFast8(2) + c#IntFast8(1)
-- output: var t1: c#IntFast8 = c#IntFast8(2)
--         var t2: c#IntFast8 = c#IntFast8(1)
--         var someVariable: c#IntFast8 = t1 + t2

module Adrian.Madgo.TAC where


import Data.Map ((!), update)

import qualified Adrian.Madgo.Env as Env
import qualified Adrian.Madgo.AST as AST
import qualified Adrian.Madgo.Inference as Inference


newContext :: Env.Context
newContext = ([Env.emptyEnv], Env.fromList [("temp_count", "0")])


createTempVariable :: AST.Expr -> Env.Context -> (String, AST.Node, Env.Context)
createTempVariable expr (envs, options) =
    let tempNameString = 't':(options ! "temp_count")
        tempDecl name =
            AST.VarDecl (AST.Name name) (Inference.inferTypeFromExpr expr) expr
        modifiedContext =
            (envs, update (\n -> Just $ show ((read n :: Integer) + 1)) "temp_count" options)
    in (tempNameString, tempDecl tempNameString, modifiedContext)


-- Applies three address code tranformation to expression.
-- Returns a tuple where first element is a new expression you need to put
-- instead of previous expression, second element is a list of
-- declarations of temporary variables and third element is a
-- modified context.
translateExpr :: AST.Expr -> Env.Context -> (AST.Expr, AST.AST, Env.Context)
translateExpr (AST.SExpr op left right) context =
    (newExpr, decls, modifiedContext)
    where
        (nameOfTemp1, declOfTemp1, _modifiedContext) = createTempVariable left context
        (nameOfTemp2, declOfTemp2, modifiedContext) = createTempVariable right _modifiedContext
        newExpr = AST.SExpr op (AST.NameInExpr nameOfTemp1) (AST.NameInExpr nameOfTemp2)
        decls = [declOfTemp1, declOfTemp2]
translateExpr expr context = (expr, [], context)


translateNode :: AST.Node -> Env.Context -> AST.AST
translateNode AST.VarDecl {AST.varName = name, AST.varType = type_, AST.varExpr = expr} context =
    let (newExpr, tempDeclarations, _) = translateExpr expr context in
    -- Declarations of temporary variables must be first, because we use them
    -- in new expression.
    tempDeclarations ++ [AST.VarDecl {
        AST.varName = name,
        AST.varType = type_,
        AST.varExpr = newExpr}]


-- Main function that applies three address code tranformation.
translateAST :: AST.AST -> AST.AST
translateAST ast = concat $ map (`translateNode` newContext) ast

File: margolith/margo/tac.py
"""
Translates expressions that are composed of several smaller ones to
three address code. Because every variable must be freed. For example, we
created a variable like this:
    `var lol: c#IntFast32 = c#IntFast32(0) + c#IntFast32(1)`,
there are no variables that point to `c#IntFast32(0)` or `c#IntFast32(1)`.
Also these transformations will help us to improve output code.
"""

import itertools

from . import astlib, layers, defs, inference
from .context import context, add_to_env, add_scope, del_scope
from .patterns import A


class TAC(layers.Layer):

    def new_tmp(self, expr):
        tmp_name = astlib.Name(
            "".join([defs.T_STRING, str(context.tmp_count)]), is_tmp=True)
        type_ = inference.infer(expr)
        context.tmp_count += 1
        node = astlib.LetDecl(tmp_name, type_, expr)
        add_to_env(node)
        return tmp_name, [node]

    def call_args(self, args_):
        args, decls = [], []
        for arg in args_:
            arg_, decls_ = self.inner_expr(arg)
            decls.extend(decls_)
            args.append(arg_)
        return args, decls

    def inner_expr(self, expr):
        if expr in A(
                astlib.FuncCall, astlib.StructCall,
                astlib.StructFuncCall):
            expr_, decls_ = self.e(expr)
            tmp, decls = self.new_tmp(expr_)
            return tmp, decls_ + decls
        if expr in A(astlib.Name, astlib.Ref, astlib.IntLiteral):
            return expr, []
        return self.new_tmp(expr)

    def _e(self, expr):
        if expr not in A(astlib.IntLiteral):
            return self.new_tmp(expr)
        return self.e(expr)

    def e(self, expr):
        if expr in A(astlib.FuncCall, astlib.StructCall):
            args, decls = self.call_args(expr.args)
            return type(expr)(expr.name, args), decls
        if expr in A(astlib.StructFuncCall):
            args, decls = self.call_args(expr.args)
            return astlib.StructFuncCall(
                expr.struct, expr.func_name, args), decls
        if expr in A(astlib.StructMember):
            if not expr.struct in A(astlib.Name):
                expr_, decls = self.inner_expr(expr.struct)
                return astlib.StructMember(expr_, expr.member), decls
        return expr, []

    def body(self, body):
        reg = TAC().get_registry()
        return list(itertools.chain.from_iterable(
            map(
                lambda stmt: list(
                    layers.transform_node(stmt, registry=reg)),
                body)))

    def _decl(self, decl):
        expr, tmp_decls = self.e(decl.expr)
        add_to_env(decl)
        yield from tmp_decls
        yield type(decl)(decl.name, decl.type_, expr)

    @layers.register(astlib.VarDecl)
    def decl(self, decl):
        yield from self._decl(decl)

    @layers.register(astlib.LetDecl)
    def ldecl(self, decl):
        yield from self._decl(decl)

    @layers.register(astlib.While)
    def while_(self, stmt):
        expr, decls = self.inner_expr(stmt.expr)
        add_scope()
        yield from decls
        yield astlib.While(
            expr, self.body(stmt.body))
        del_scope()

    @layers.register(astlib.Assignment)
    def assignment(self, stmt):
        expr, decls = self.e(stmt.expr)
        yield from decls
        yield astlib.Assignment(
            stmt.variable, stmt.op, expr)

    @layers.register(astlib.Return)
    def return_(self, stmt):
        expr, decls = self.inner_expr(stmt.expr)
        yield from decls
        yield astlib.Return(expr)

    @layers.register(astlib.FuncDecl)
    def func_decl(self, stmt):
        add_to_env(stmt)
        add_scope()
        yield astlib.FuncDecl(
            stmt.name, stmt.args, stmt.rettype, self.body(stmt.body))
        del_scope()

    @layers.register(astlib.StructFuncDecl)
    def struct_func_decl(self, stmt):
        add_to_env(stmt)
        add_scope()
        yield astlib.StructFuncDecl(
            stmt.struct, stmt.func, stmt.args,
            stmt.rettype, self.body(stmt.body))
        del_scope()

    @layers.register(astlib.StructDecl)
    def struct_decl(self, stmt):
        add_to_env(stmt)
        add_scope()
        yield astlib.StructDecl(
            stmt.name, stmt.var_types, self.body(stmt.body))
        del_scope()

    @layers.register(astlib.ADTDecl)
    def adt_decl(self, stmt):
        add_to_env(stmt)
        add_scope()
        yield stmt
        del_scope()
